application-analytics-ui/src/main/resources/Analytics/Code/Macros/countryJSON.xml

```diff
@@ -0,0 +1,106 @@
+<?xml version="1.1" encoding="UTF-8"?>
+
+<!--
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+-->
+
+<xwikidoc version="1.5" reference="Analytics.Code.Macros.countryJSON" locale="">
+  <web>Analytics.Code.Macros</web>
+  <name>countryJSON</name>
+  <language/>
+  <defaultLanguage/>
+  <translation>0</translation>
+  <creator>xwiki:XWiki.Admin</creator>
+  <parent>Analytics.Code.Macros.Country</parent>
+  <author>xwiki:XWiki.Admin</author>
+  <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
+  <version>1.1</version>
+  <title>countryJSON</title>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>true</hidden>
+  <content>{{include reference='Analytics.Code.Macros.VelocityMacros'/}}
+{{velocity}}
+#if ($xcontext.action == 'get')
+ #analytics_livedataPagination()
+  ## Mapping between the livedata columns and the Matomo fields.
+  #set ($matomoKeysMapping = {
+    'country' : 'label',
+    'visits': 'nb_visits',
+    'userActions':'nb_actions',
+    'timeSpent':'sum_visit_length'
+  })
+  #set ($sort = 'nb_visits')
+  #set ($order = 'desc')
+  #if ($request.sort)
+    #set ($sort = $matomoKeysMapping.get($request.sort))
+    #set ($order = $request.dir)
+  #end
+
+  #analytics_readFiltersFromRequest($matomoKeysMapping)
+  #set ($parameters = {
+    'period' : $request.period,
+    'date' : $request.date,
+    'module' : 'API',
+    'method' : $request.apiMethod,
+    'format' : 'json',
+    'filter_limit' : $request.limitEntries,
+    'filter_sort_column' : $sort,
+    'filter_sort_order' : $order,
+    'expanded' : '1',
+    'flat' : '1'
+  })
+
+  ## Get the JSON response from Matomo.
+  #set ($analyticsResult = $services.analytics.makeRequest($parameters, $filters, 'MostViewedPages'))
```

Comment:
Looks like you get all the results and do the pagination afterwards. Is there no way for the Matomo API to return a paginated result?

Reply:
There is a way to implement pagination directly with Matomo. However, the issue lies in the Live Data, as it requires the total number of rows. Unfortunately, Matomo does not have an endpoint to return the number of rows a request would yield.